text/0000-ghost-busting.md

```diff
@@ -0,0 +1,900 @@
+
+3. `PhantomData<T>` is often used when building abstractions that really own
+   `T` in memory such as `pub struct RawTable<K, V>` in the standard library.
+   And if these data structures do not really own a value of type `T`, they
+   often can or will produce owned values of type `T` as in the case of
+   iterators (`Empty<T>`), futures, or strategies to randomly generate a value.
+   This means that owned `T`s are reachable from the type which contained the
+   `PhantomData<T>`.
+
+## Alternative: Add `PhantomData<T>` to the prelude
+
+By adding `PhantomData<T>` to the prelude, we can at least get rid of the step
+that you have to import `PhantomData`. However, this does not solve the problem
+of having to add a marker type or to pass around an expression `PhantomData` in
+literals.
+
+## Alternative: Allow filling unspecified fields with `Default`
+
+Simply put, this would mean that a struct literal such as `Foo { .. }` would
+be a shorthand for the FRU `Foo { .. Default::default()`. This makes dealing
+with `PhantomData` more ergonomic since expressions involving those do not
+have to write out `field: PhantomData`. However, this does not solve the issue
+of pain when dealing with type definitions.
+
+[RFC 1806]: https://github.com/rust-lang/rfcs/pull/1806
+
+A slightly modified version of this was proposed in the now postponed [RFC 1806].
+
+A further problem this presents is that given an API:
+
+```rust
+mod global_lock {
+    pub fn take() -> Locked { ... }
+
+    pub struct Locked { private: () }
+
+    impl Drop for Locked { ... }
+
+    pub fn do_stuff(_: Locked) { ... }
+}
+```
+
+the type `Locked` is a token for a proof of work even if it is only a
+zero-sized-type (ZST). Being able to produce a `Locked` out of thin air
+can break invariants. Solving this will mean that only some ZSTs must be
+able to be elided and not others; this added complexity is better solved
+by dedicated UX for phantoms.
+
+## Alternative: `#[phantom]` attributes on fields
+
+This alternative would allow a user to define `Id<A, B>` as:
+
+```rust
+struct Id<A, B> {
+    #[phantom]
+    _marker: (fn(A) -> A, fn(B) -> B),
+}
+```
+
+This has the benefit that privacy is controlled via normal fields and does
+not introduce any new surface syntax. However, this idea has a few problems:
+
+1. You have to invent a name `_marker` for what should be unnameable.
+2. It is not at all obvious that you can't access a value of type `T` with
+   `self._marker`. A reasonable reader can easily assume this. If we permit
+   such an attribute - what would the type of `_marker` be? If it is
+   `PhantomData<T>`, then the mental model of types is instead complicated
+   and a library type becomes even more magical.
+
+## Alternative: `#[phantom(T)]` attributes on type parameters
+
+This alternative would allow a user to define `Id<A, B>` as:
+
+```rust
+struct Id<#[phantom(fn(A) -> A)] A, #[phantom(fn(A) -> A)] B> {
```

Comment:
I'm not sure if this should be `fn(B) -> B`, but it does illustrate a huge drawback with parameter attributes 😏

Reply:
Hah; you are on fire today =) Thanks a lot for spotting these!